src/docker.c

```diff
@@ -0,0 +1,1137 @@
+  stats->cpu_stats = (cpu_stats_t *) calloc(1, sizeof(cpu_stats_t));
+  if (stats->cpu_stats == NULL) {
+    ERROR("docker: get_cpu_stats: malloc failed!");
+  }
+  yajl_val cpu_node = yajl_tree_get(node, CPU_PATH, yajl_t_object);
+  if (!YAJL_IS_OBJECT(cpu_node)) {
+    ERROR("docker: JSON Error. \n");
+    sfree(stats->cpu_stats);
+    return;
+  }
+
+  unsigned long *result_ptrs[] = {
+    &(stats->cpu_stats->system_cpu_usage),
+  };
+
+  assert(STATIC_ARRAY_SIZE(CPU_KEYS) == STATIC_ARRAY_SIZE(result_ptrs));
+  for (int i = 0; i < STATIC_ARRAY_SIZE(CPU_KEYS); i++) {
+    extract_value(cpu_node, CPU_KEYS[i], result_ptrs[i]);
+  }
+  int len = extract_arr_value(cpu_node, "cpu_usage,percpu_usage",
+      &(stats->cpu_stats->percpu_usage));
+  stats->cpu_stats->num_cpus = (unsigned long) len;
+  stats->cpu_stats->percpu_idle =
+      (unsigned long *) calloc(len, sizeof(unsigned long));
+  stats->cpu_stats->percpu_percent_used = (double *) calloc(len, sizeof(double));
+  stats->cpu_stats->percpu_percent_idle = (double *) calloc(len, sizeof(double));
+  compute_cpu_stats(stats->cpu_stats, container_id);
+}
+
+// Retrieves Memory statistics from the STATS API response
+static void get_memory_stats(yajl_val node, container_stats_t * stats) {
+  stats->memory_stats = (memory_stats_t *) calloc(1, sizeof(memory_stats_t));
+  if (stats->memory_stats == NULL) {
+    ERROR("docker: get_memory_stats: malloc failed!");
+  }
+  yajl_val memory_node = yajl_tree_get(node, MEMORY_PATH, yajl_t_object);
+  if (!YAJL_IS_OBJECT(memory_node)) {
+    ERROR("docker: JSON Error. \n");
+    sfree(stats->memory_stats);
+    return;
+  }
+
+  unsigned long *result_ptrs[] = {
+    &(stats->memory_stats->usage),
+    &(stats->memory_stats->limit),
+  };
+
+  assert(STATIC_ARRAY_SIZE(MEMORY_RESPONSE_KEYS) ==
+         STATIC_ARRAY_SIZE(result_ptrs));
+  for (int i = 0; i < STATIC_ARRAY_SIZE(MEMORY_RESPONSE_KEYS); i++) {
+    extract_value(memory_node, MEMORY_RESPONSE_KEYS[i], result_ptrs[i]);
+  }
+  compute_memory_stats(stats->memory_stats);
+}
+
+// Extracts stats for a given interface.
+static void get_interface_stats(yajl_val interface_node,
+    interface_stats_t * stats) {
+  //Network Stats
+  unsigned long *rx_result_ptrs[] = {
+    &(stats->rx_bytes),
+    &(stats->rx_packets),
+    &(stats->rx_errors),
+  };
+
+  unsigned long *tx_result_ptrs[] = {
+    &(stats->tx_bytes),
+    &(stats->tx_packets),
+    &(stats->tx_errors),
+  };
+
+  assert(STATIC_ARRAY_SIZE(INTERFACE_KEYS) == STATIC_ARRAY_SIZE(rx_result_ptrs));
+  assert(STATIC_ARRAY_SIZE(INTERFACE_KEYS) == STATIC_ARRAY_SIZE(tx_result_ptrs));
+  for (int i = 0; i < STATIC_ARRAY_SIZE(INTERFACE_KEYS); i++) {
+    int len = strlen(INTERFACE_KEYS[i]) + 4; //tx_{bytes}\0
+    char *tx_key = (char *) calloc(len, sizeof(char));
+    char *rx_key = (char *) calloc(len, sizeof(char));
+    if (tx_key == NULL || rx_key == NULL) {
+      ERROR("docker: get_metrics_for_container: malloc failed.\n");
+    }
```

Comment:
Why not abort / return to the caller here instead of doing more (unnecessary) processing?